### PR TITLE
Implement the Diagnostics gRPC service

### PR DESCRIPTION
**Pull request by lonng**
Signed-off-by: Lonng <heng@lonng.org>

###  What have you changed?

Initial implementation of diagnostics log service, which is part of https://github.com/pingcap/tidb/pull/13481

**Known issue**

This PR is implemented using synchronous IO and refactored to asynchronous IO in the next PR.

###  What is the type of the changes?

Pick one of the following and delete the others:

- New feature (a change which adds functionality)

###  How is the PR tested?
Please select the tests that you ran to verify your changes:
- Unit test
- Manual test (add detailed scripts or steps below)

### CODE REVIEW

File: src/server/service/diagnostics.rs [hunk 1]
```diff
@@ -0,0 +1,690 @@
...
+                None => {
+                    return Err(Error::SearchError(format!(
+                        "Illegal parent dir: {:?}",
+                        log_path
+                    )))
+                }
+            };
+
+            let mut search_files = vec![];
+            for entry in read_dir(log_dir)? {
```
Comment (lonng) [inline src/server/service/diagnostics.rs]: For reviewer: This PR is temporarily implemented using synchronous IO, which will later refactor asynchronous IO (implemented by me or by community members).  The reason is that we need to implement an asynchronous version of `rev_lines` dependency. Consider the modification of this PR is very huge, so split into two parts.

File: src/server/service/diagnostics.rs [hunk 2]
```diff
@@ -0,0 +1,682 @@
+use futures::Future;
```
Comment (overvenus) [inline src/server/service/diagnostics.rs]: License header？

File: src/server/service/diagnostics.rs [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+        &mut self,
+        _ctx: RpcContext<'_>,
+        _req: ServerInfoRequest,
+        _sink: UnarySink<ServerInfoResponse>,
+    ) {
+        unimplemented!()
+    }
+}
+
+mod log {
```
Comment (overvenus) [inline src/server/service/diagnostics.rs]: Could you move it to a dedicated mod folder? Eg, `src/server/diagnostics/log.rs`.

Comment (lonng) [reply src/server/service/diagnostics.rs]: I think it's better to keep put the log module in this file (other services in the `service` module is a single file), but we can take this into account if someday the log sub module growth.

File: src/server/service/diagnostics.rs (hunk 2 above)
Comment (lonng) [reply src/server/service/diagnostics.rs]: addressed

File: src/server/service/diagnostics.rs [hunk 4]
```diff
@@ -0,0 +1,684 @@
...
+        &mut self,
+        ctx: RpcContext<'_>,
+        req: SearchLogRequest,
+        sink: UnarySink<SearchLogResponse>,
+    ) {
+        let future = log::search(&self.log_file, req)
+            .map_err(|err| Error::Other(box_err!("Search log error: {:?}", err)))
+            .and_then(|res| sink.success(res).map_err(Error::from))
+            .map_err(|e| debug!("Diagnostics rpc failed"; "err" => ?e));
+        ctx.spawn(future);
```
Comment (hicqu) [inline src/server/service/diagnostics.rs]: The service is registered into same environment with `kv::Service`. So they share grpc threads. Is it better to move the logic into a thread pool? I think sharing the pool with `Debug` is ok.

File: src/server/service/diagnostics.rs [hunk 5]
```diff
@@ -0,0 +1,684 @@
...
+                };
+
+                let (file_start_time, file_end_time) = match parse_time_range(&file) {
+                    Ok((file_start_time, file_end_time)) => (file_start_time, file_end_time),
+                    Err(_) => continue,
+                };
+                if (begin_time <= file_start_time && end_time >= file_start_time)
+                    || (end_time >= file_end_time && begin_time <= file_end_time)
+                {
+                    if let Err(err) = file.seek(SeekFrom::Start(0)) {
```
Comment (hicqu) [inline src/server/service/diagnostics.rs]: It's unnecessary.

Comment (lonng) [reply src/server/service/diagnostics.rs]: It' necessary because the same file descriptor will be reused.

**Pull request merged**